contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,215 @@
+    exit or cancel) followed by a time in seconds for the giveaway
+    to run for. Full syntax is shown below:
+    @giveaway command giveaway_name giveaway_time
+
+    Possible commands:
+    init - initialize a giveaway (only one can be running at a time)
+    enter - enter a giveaway (a giveaway must be running to use this command)
+    exit - leave a giveaway (must be partaking in an active giveaway for this
+    command to work
+    cancel - cancel a giveaway (can only be run by the person who first
+    initialised the giveaway
+
+    This plugin was created by Daniel O'Brien
+    as part of Google Code-In 2016-2017
+    '''
+    def __init__(self):
+        self.giveaway_array = []
+        self.giveaway_on = False
+        self.giveaway_name = ""
+        self.giveaway_time = 0
+        self.giveaway_creator = ""
+        self.giveaway_start_time = 0
+        self.giveaway_group = ""
+        self.giveaway_subject = ""
+        self.giveaway_client = ""
+        self.giveaway_timeMins = 0
+        self.giveaway_stage = 0
+        self.one_minute_warning = False
+        self.giveaway_winner = ""
+        self.giveaway_winner_contact_address = []
+        self.winner_index = 0
+        self.all_warnings_fired = False
+
+    def time_update(self, new_content):
+        if self.giveaway_on:
+            self.giveaway_client.send_message(dict(
+                type='stream',
+                to=self.giveawayStream,
+                subject=self.giveaway_subject,
+                content=new_content,
+            ))
+            threading.Timer(6.0, self.check_time).start()
+
+    def time_warning(self):
+        if self.giveaway_stage == 0:
+            self.numerator = 1
+            self.denominator = 4
+        elif self.giveaway_stage == 1:
+            self.numerator = 1
+            self.denominator = 2
+        elif self.giveaway_stage == 2:
+            self.numerator = 3
+            self.denominator = 4
+        elif self.giveaway_stage == 0:
+            self.all_warnings_fired = True
+        if time.time() > ((self.giveaway_start_time + self.giveaway_time)-60):
+            if self.giveaway_timeMins > 1.9 and not self.one_minute_warning:
+                new_content = 'Giveaway ends in 1 minute'
+                self.time_update(new_content)
+                self.one_minute_warning = True
+        elif time.time() > (self.giveaway_start_time + (old_div(self.giveaway_time, self.denominator)*self.numerator)):
+            if not self.all_warnings_fired:
+                new_content = ('{}/{} of giveaway time elapsed, {} minutes remaining'
+                               .format(self.numerator, self.denominator, str((old_div(self.giveaway_timeMins, 4))*3)))
+                self.time_update(new_content)
+                self.giveaway_stage = self.giveaway_stage + 1
+
+    def check_time(self):
+        if time.time() > (self.giveaway_start_time + self.giveaway_time):
+            print("Giveaway ended")
```

Comment:
For long running processes, it's better to use `logging`, in production at least.